feat(lsp)!: reimplement LSP document highlight with the new API style

ofseed opened the pull request:
## Problems
`vim.lsp.buf.document_highlight` was implemented quite a long time ago. Its API style differs from our latest APIs, such as `vim.lsp.inlay_hint`. I believe it's time to give it an overhaul.

## Solution
* Reimplement `textDocument/documentHighlight` as a standalone module, offering an API similar to `vim.lsp.document_highlight.enable()`.
* Additionally, `vim.lsp.document_highlight.jump()` for jumping between highlights.

[document_highlight.webm](https://github.com/user-attachments/assets/67755d5b-5084-4171-96e3-9fa2bc9dce5e)

justinmk commented:
> but like `vim.lsp.inlay_hint`, this API needs to manage an independent `bufstate`. We can explore solutions such as [Lua: store metatables on vim.b/vim.w/vim.t/vim.v scopes #12544](https://github.com/neovim/neovim/issues/12544) or refactor LSP `bufstate` management into a standalone module (since the lifecycle of these variables is tied to events like `LspAttach` and `LspDetach`)

If there is a common pattern amongst these lsp modules then yes they should use a single common lsp-internal-only module that implements this. Several of these lsp modules have implemented their own long-lived "lifecycle" maps, and that is not something that should be reinvented in each.

glepnir commented on runtime/lua/vim/lsp/document_highlight.lua:
```diff
@@ -0,0 +1,189 @@
+  vim.lsp.buf_request(bufnr, ms.textDocument_documentHighlight, params)
+end
+
+api.nvim_create_autocmd('CursorMoved', {
```
It is better to avoid CursorMoved

ofseed replied:
Because `textDocument/documentHighlight` needs to get the cursor's position in the document (it is used for fuzzy matching the references of the symbol under the cursor), I don't know know to avoid it. Since LSP requests are sent asynchronously, and we are not handling its return value here, I think it shouldn't cause too much of a performance issue, but I intended to implement debouncing here. If there is a better way, please let me know.

glepnir replied:
You  to send the current cursor position to the server. What happens when you move quickly like holding jk is that the refresh function keeps sending requests. no processing here.

Use a timer here and store the last request. When it is triggered again and still within the timeout, you need to cancel the previous request and start a new timer. When the callback function is called, you also need to determine whether the cursor position is the same as the previous cursor position. This is no different from CursorHold. Users can adjust it according to updatetime

ofseed replied:
> there send the current cursor position to the server. What happens when you move quickly like holding jk is that the refresh function keeps sending requests. no processing here.

As mentioned earlier, this is exactly my plan (I haven't optimized the performance yet; it's currently just a rough implementation, but I intend to do so, just like the todo list above).

I'm not sure if `CursorHold` is a good idea, It shouldn't be used for this purpose because we do want to trigger it after the cursor is moved. The default `updatetime` is too long for debouncing, and it's a global option that shouldn't be changed for one specific purpose.

Later when this PR is ready to merge, I will request a review.

glepnir replied:
Yes CursorHold is not a good idea here either. CursorMoved is also sensitive. At least you need to handle some of the behaviors I said.

I'm not sure if we need this. Plugins need a certain amount of space. Not all in core.

lewis6991 replied:
I don't use a plugin for this. I have this in my config:

```lua
do -- textDocument/documentHighlight
  local method = 'textDocument/documentHighlight'

  api.nvim_create_autocmd({ 'FocusGained', 'WinEnter', 'BufEnter', 'CursorMoved', 'CursorHold', 'CursorHoldI' }, {
    callback = debounce(200, function(args)
      lsp.buf.clear_references()
      local bufnr = args.buf --- @type integer
      for _, client in ipairs(lsp.get_clients({ bufnr = bufnr })) do
        if client.supports_method(method, { bufnr = bufnr }) then
          local params = lsp.util.make_position_params()
          client.request(method, params, nil, bufnr)
        end
      end
    end)
  })

  api.nvim_create_autocmd({ 'FocusLost', 'WinLeave', 'BufLeave' }, {
    callback = lsp.buf.clear_references
  })
end
```

It's not perfect, but works well enough.

clason replied:
Tangential, but it would make sense to refactor this (which already exists somewhere in the changetracking module) and expose this. This is exactly the sort of short but non-trivial snippet that having in our stdlib would be a significant QoL improvement for the ecosystem.

(Same with `memoize` and async, *ceterum censeo*.)

lewis6991 replied:
I think `schedule` and `schedule_wrap` work well as separate functions as they are both used plenty and they have different return types (`nil` vs `function`). It's similar to how Lua has `coroutine.wrap` vs `coroutine.create` which return `function` and `thread` respectively.

ofseed replied:
In this case, I believe the closest-to-perfect solution would require resolving https://github.com/neovim/neovim/issues/27509. Other methods seem more like workarounds.

ofseed replied:
(Mark as resolved because the debounce logic has been added.)

glepnir commented on runtime/lua/vim/lsp/document_highlight.lua:
```diff
@@ -0,0 +1,189 @@
+          local col = highlight.range['start'].character
+          local end_row = highlight.range['end'].line
+          local end_col = math.min(highlight.range['end'].character, max_col)
+          vim.api.nvim_buf_set_extmark(bufnr, namespace, row, col, {
```
```suggestion
          api.nvim_buf_set_extmark(bufnr, namespace, row, col, {
```

mfussenegger commented on runtime/lua/vim/lsp/document_highlight.lua:
```diff
@@ -0,0 +1,323 @@
+})
+api.nvim_create_autocmd('LspNotify', {
+  group = augroup,
+  callback = function(args)
+    ---@type integer
+    local bufnr = args.buf
+    local client = assert(vim.lsp.get_client_by_id(args.data.client_id))
+    if
+      client.supports_method(ms.textDocument_documentHighlight, { bufnr = bufnr })
+      and (
+        args.data.method == ms.textDocument_didChange
+        or args.data.method == ms.textDocument_didOpen
+      )
+    then
+      debounced_refresh(bufnr)
+    end
+  end,
+})
```
this will have the side effect that requests triggered by other plugins or custom user functions would also trigger a new refresh - I'm not sure if that's desired?

Why is this additional LspNotify autocmd necessary?

mfussenegger commented on runtime/lua/vim/lsp/document_highlight.lua:
```diff
@@ -0,0 +1,323 @@
+
+---@param kind lsp.DocumentHighlightKind
+---@return string
+local function kind_to_hl_group(kind)
+  local base = 'LspReference'
+  if kind == 2 then
+    return string.format('%sRead', base)
+  elseif kind == 3 then
+    return string.format('%sWrite', base)
+  else -- kind == 1 also the default
+    return string.format('%sText', base)
+  end
+end
```
The old implementation used a map for this which seems more straightforward. Any reason to change this?

ofseed replied:
For `DocumentHighlight`, the standard specifies that `kind` is optional. This approach may be more straightforward than a map, as it better represents a fallback.

https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#documentHighlightKind

lewis6991 replied:
As `base` is a constant I would inline it and remove all the `string.format`'s.

ofseed replied:
I initially planned to change this number because `LspReference` is ambiguous. However, I couldn't think of a better name, as `LspDocumentHighlight` is too long and unnecessary. I can change this.

mfussenegger commented on runtime/lua/vim/lsp/document_highlight.lua:
```diff
@@ -0,0 +1,323 @@
+    return
+  end
+
+  local params = util.make_position_params()
+  vim.lsp.buf_request(bufnr, ms.textDocument_documentHighlight, params)
```
Params would need to be created per client with `client.offset_encoding`

ofseed replied:
Thank you for your reminder. I haven't completed the work related to offset_encoding yet. I will look into this.

mfussenegger reviewed (commented):
Looking at the previous implementation it's not clear to me if going with the decoration provider is an improvement in this case. Might be simpler to keep the old pattern, but build enable/disable convenience around it?
Or maybe we should just setup some autocmds by default with no new APIs similar to what Lewis posted and tell users they can opt out by deleting a specific augroup?
Feels to me that the decision when to show the highlights is something users might want to control. This has somehow different semantics than something like semantic tokens given that it's position dependant

Looks like it's also sacrificing a bit of flexibility (e.g. highlighting multiple different symbols at the same time)

ofseed commented:
> Feels to me that the decision when to show the highlights is something users might want to control. This has somehow different semantics than something like semantic tokens given that it's position dependant

There is indeed a certain issue with flexibility, and this is the key to the problem, but I have long-term goals regarding this. I have previously noticed that a considerable number of people want to customize the display of inlay hints, as evidenced by the strong reactions and numerous approvals in https://github.com/neovim/neovim/issues/28261, but we did not find a solution.

In LSP, the three methods have similarities, besides the `textDocument/documentHighlight` that is being attempted to change here, they also include `textDocument/inlayHint` and `textDocument/codeLens`. If they are all implemented by the decoration provider. then we have enough reason to provide a unified interface, something like `on_decorate`, for all three methods. This is one of my long-term goals.

Although many people want to define the way highlights are displayed, I don't believe there are as many people who want to manage `bufstate`, which is the management of how this data to be displayed is updated, this is just cumbersome work. This part is handled by our traditional `lsp-handler`, which is easy for users to modify, that's good, but modifying them doesn't make sense for the three LSP methods I mentioned earlier, so users either have to use built-in features that cannot be customized or completely take over the implementation of these requests, which is why I believe there is a necessity for `on_decorate`.

As for the management of bufstate, I also have some ideas. Similar to what @justinmk mentioned earlier, I plan to refactor an internal-only `vim.lsp._state` module to manage the lifecycle of bufstate required for `textDocument/documentHighlight`, `textDocument/inlayHint`, and `textDocument/codeLens`. This will greatly allow us to reuse code while improving flexibility.

lewis6991 commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -1679,60 +1678,6 @@ function M.open_floating_preview(contents, syntax, opts)
-  --- Removes document highlights from a buffer.
-  ---
-  ---@param bufnr integer? Buffer id
-  function M.buf_clear_references(bufnr)
```
These can't be removed. They need to be deprecated first.

lewis6991 commented on runtime/lua/vim/lsp/document_highlight.lua:
```diff
@@ -0,0 +1,350 @@
+    end
+  end
+end
+api.nvim_create_autocmd('LspAttach', {
```
I don't think these should be created at `require` time. These should either be placed somewhere in `plugin`, or setup via some other function. Perhaps `lsp.start()`?

ofseed replied:
Makes sense, but this part is likely temporary. I will add a dedicated PR for managing bufstate after a while(for this module and `inaly_hint.lua`), and the logic is based on this part, so these autocmds will be removed from this module regardless. And `inlay_hint.lua` has a similar logic, so maybe in follow-up PRs? Is there any vulnerability here?

lewis6991 commented on runtime/lua/vim/lsp/document_highlight.lua:
```diff
@@ -0,0 +1,350 @@
+      return false
+    end
+
+    if bufstates[bufnr].enabled == nil then
+      return globalstate.enabled
+    else
+      return bufstates[bufnr].enabled
+    end
```
```suggestion
    if bufstate.enabled == nil then
      return globalstate.enabled
    else
      return bufstate.enabled
    end
```

lewis6991 commented on runtime/lua/vim/lsp/document_highlight.lua:
```diff
@@ -0,0 +1,350 @@
+  local bufnr = filter.bufnr
+  if bufnr == nil then
+    return globalstate.enabled
+  else
```
```suggestion
  end
```

Don't nest after a early return.

ofseed commented:
I have completed the minimum target functionality, please feel free to review or test this PR. I have not written a standalone test module for nvim yet, so adding tests will take some time.

mfussenegger commented:
> I have previously noticed that a considerable number of people want to customize the display of inlay hints, as evidenced by the strong reactions and numerous approvals in #28261, but we did not find a solution.

It also wasn't a goal to find one that makes customization easier than what's already possible.

If we wanted to find a way to customize the highlighting of all inlay_hints, document_highlights and code_lens then I'm not sure if this should depend on all of them having a decoration provider implementation. As that's also a bit implementation specific and I'm not sure if that aspect should leak to users.

I also question a bit if that should be a goal, given that there aren't that many options on how to display them within the constraints from the specification.

>  If they are all implemented by the decoration provider. then we have enough reason to provide a unified interface, something like on_decorate, for all three methods. This is one of my long-term goals.

Can you give a bit of a peek how this would look like - and also how a supposed `lsp._state` would look like?

Both codelens and inlay-hints have the resolve request addition, that's one of the reasons they need dedicated state management. Document highlights don't have that requirement and what made the implementation so far a lot simpler. The autocmds and highlights were sufficient.


Don't get me wrong, I'm not opposed to changing stuff around to enable new functionality or make the setup easier. But so far this seems to me a bit solution oriented and I'm missing why we'd incur churn for users when in the end some might even loose functionality they currently make use of and the implementation is more complex than before.

justinmk commented:
> As for the management of bufstate, I also have some ideas. Similar to what @justinmk mentioned earlier, I plan to refactor an internal-only `vim.lsp._state` module to manage the lifecycle of bufstate

Wouldn't it make sense to start with that? There are at least 2 lsp modules that could be refactored to use this. That is a more targeted change, that is also useful for future lsp work.

ofseed commented:
> Both codelens and inlay-hints have the resolve request addition, that's one of the reasons they need dedicated state management and why a decoration provider makes sense to me as it allows to restrict resolving items to the active view-port which helps performance. Document highlights don't have that requirement and that made the implementation so far a lot simpler. The autocmds and highlights were sufficient.

So I think our disagreement lies in whether to manage the state of document highlights? I have two reasons for supporting this.

Functionally, it allows us to provide some more advanced features, such as jumping to the previous highlighted location or jumping to the next highlighted location, which was in the previous to-do list, but I think it might be an independent PR, so I deleted it. Explaining why this feature is useful and why it needs to be implemented could take quite a long time; I can explain it in another PR. It is important to note that this is not functionally overlapping with `vim.lsp.buf.references`, as it provides fuzzy matching, and if the state is saved, its implementation would be quite simple.

For complexity, the complexity of the current implementation comes from the management of bufstate. If it is not managed by it, all implementations will become very simple. In fact, I have already implemented my ideas and logic for managing them in this PR.

> and also how a supposed `lsp._state` would look like?

> Wouldn't it make sense to start with that (separate PR)? 

OK, I will freeze this PR now and start with this.

mfussenegger commented:
> So I think our disagreement lies in whether to manage the state of document highlights? 

Mostly. autocmds are implicit state of enabled/disabled and the decoration provider doesn't bring any benefits over setting the highlights in the callback given that highlights tend to be few. 

> Functionally, it allows us to provide some more advanced features, such as jumping to the previous highlighted location or jumping to the next highlighted location, which was in the previous to-do list, but I think it might be an independent PR, so I deleted it

This doesn't need state management either but can be done by firing new requests. I did that in nvim-overfly:

https://github.com/mfussenegger/nvim-overfly/blob/48df0c67ff59d0550ff306579e8aa4e9193b665f/lua/overfly.lua#L5-L60

clason commented:
> such as jumping to the previous highlighted location or jumping to the next highlighted location

That is indeed a compelling use case (but of course not at _any_ cost). In general, it's better to work "inside out", starting with the low-level functionality (but providing a roadmap either in the PR description or in a tracking issue) and moving to the user-facing features rather than doing "vertical slices".

ofseed commented:
> This doesn't need state management either but can be done by firing new requests. I did that in nvim-overfly:

This is also an example of why I think should be implemented by storing. If it is sync, it will block; if it is async, the cursor may be unexpectedly moved after the user moves. Using state to implement both cases can avoid this, and most importantly, it ensures what you see is what you get.

> but of course not at any cost

And this is the key, costless except for the state, while state management has other more important reasons to be implemented.

> starting with the low-level functionality (but providing a roadmap either in the PR description or in a tracking issue) and moving to the user-facing features rather than doing "vertical slices".

Make sense.

mfussenegger commented:
> If it is sync, it will block; if it is async, the cursor may be unexpectedly moved after the user moves

At least with luals, pylsp and eclipse.jdt.ls the move happens pretty much instant in my experience with overfly. No blocking noticable.
I also often use it before the highlight based on a CursorHold kicks in, but if you wanted to ensure that what you see is what you get: the state would also already be there in the form of extmarks. (`:Inspect` shows the document highlights) - One could also use them to jump to the next extmark with a given namespace/highlight group. That could even be generalized if desired.
That said, to me jumping based on a deferred highlight state would be a Verschlimmbesserung and I'd stick with the overfly implementation as it doesn't have the debounce latency.

To summarize my points:

- The state already exists as aucmds or extmarks and we'd just be deduplicating it - that is required to some degree for inlay hints and code-lens due to the resolve logic, but not for document highlights.

- If the main problem is that it's too much effort to enable document highlight we could either:
    - consider enabling it by default (disable: via changing capabilities or clearing a augroup)
    - have a .enable()/.is_enabled based on autocmd/augroups without additional state

- If the main motivation is the jump/move:
    - Could be based on extmarks
    - Could be based on dedicated request
    - Or something that aligns with future vim.jump plans (e.g. `vim.jump(vim.lsp.doc_hl.get({ direction = 1 })[1])`)

ofseed commented:
I must admit that it is still too early to discuss features like `vim.lsp.document_highlight.jump`. My starting point is that providing this feature is simple, but right now I have no way to confirm this because the preliminary work planned has not been completed. Once these goals are achieved, I will try to answer the questions raised by @mfussenegger.

My current planned roadmap:

- [ ] `vim.lsp._state`, which is confirmed to be necessary, will provide it in a PR that refactors `inlay_hint.lua`.
- [ ] Add `enable` and `is_enabled` for `vim.lsp.codelens`. Based on previous discussions, I think we agreed that `enable` and `is_enable` is an acceptable interface and codelens is suitable for utilizing `nvim_set_decoration_provider`, right? This will be a separate PR.
- [ ] Complete this PR. May implement `jump` here to check whether it's worth.
- [ ] Implement `textDocument/foldingRange`, provide something like `vim.lsp.foldexpr`. Surprise! I also have plans to implement new features, and this feature's implementation should based on `vim.lsp._state`.
- [ ] Explore `vim.lsp.config`. One major feature that `textDocument/foldingRange` can bring is automatic folding of imports. We definitely need an interface to configure it if the feature was implemented, this can be done in a separate module, but I think if we have a lot of `enable` and `is_enable`, a unified interface should be valuable.

I believe the goals to be implemented mentioned above won't be too complex, while also adhering to the code standards here. I hope nvim can maintain high extensibility while also being more out of the box, which is why I am not creating a standalone plugin or just implementing it casually in my dotfiles. Of course, this is all contingent on the agreement of you maintainers.

igorlfs commented:
> It's not clear if that is a discoverability problem or if people regard installing a plugin as easier than setting up two autocmds.

As a regular user, I'm fine with setting up the autocmds. However, I still prefer using a plugin. IMO, the current (from master) suggested setup has two main limitations:
1. Since it uses the `CursorHold` event, if the cursor moves too fast, multiple regions may be highlighted (which has been addressed in this discussion)
2. Since the other autocmd uses the `CursorMoved` event, references are often cleared unnecessarily (i.e., when moving within the same word)

ofseed commented:
@igorlfs This is not our point of divergence. I believe we have reached a consensus that it is necessary to provide an `enable()` interface to enable directly, and our current discussion focuses on the implementation method.

I force-pushed with `vim.lsp.document_highlight.jump` implemented.

> I also often use it before the highlight based on a CursorHold kicks in, but if you wanted to ensure that what you see is what you get: the state would also already be there in the form of extmarks. (`:Inspect` shows the document highlights) - One could also use them to jump to the next extmark with a given namespace/highlight group. That could even be generalized if desired.

This potential issue has been resolved in there (it's not complicated). I implemented it because I currently need this feature, and I am still trying to implement `vim.lsp._state`.

glepnir commented on runtime/lua/vim/lsp/document_highlight.lua:
```diff
@@ -0,0 +1,486 @@
+          if row == pos[1] then
+            if highlight.range['start'].character > pos[2] then
+              pos = { highlight.range['start'].line, highlight.range['start'].character }
+              goto continue
```
dont' use `goto`, it's extension from luajit not PUC lua 5.1

glepnir commented on runtime/lua/vim/lsp/document_highlight.lua:
```diff
@@ -0,0 +1,502 @@
+---@param enable? boolean
+---@param filter? vim.lsp.document_highlight.enable.Filter
+function M.enable(enable, filter)
+  vim.validate('enable', enable, 'boolean', true)
+  vim.validate('filter', filter, 'table', true)
```
just call vim.validate once, `enable` and `filter` can be in a table

clason replied:
No, this would revert a recent PR from @lewis6991. This was deliberate. See #30855.

ofseed replied:
Yes, it was suggested by lewis [there](https://github.com/neovim/neovim/pull/30841#discussion_r1806116877)

glepnir replied:
np i just see the code without see more context.

glepnir commented on runtime/lua/vim/lsp/document_highlight.lua:
```diff
@@ -0,0 +1,502 @@
+      return
+    end
+
+    if enable == globalstate.enabled then
+      bufstate.enabled = nil
+    else
+      bufstate.enabled = enable
+    end
```
```suggestion
    bufstate.enabled = enable ~= globalstate.enabled and enable or nil
```

ofseed replied:
This is not correct. `and ... or` is not equivalent to the regular ternary operation. In `condition and expr1 or expr2`, if the value of `expr1` can be `nil` or `false`, regardless of the result of `condition`, the obtained value will always be `expr2`. This is what I deliberately want to avoid.

glepnir replied:
nil trap should be avoided. enable always has a boolean available here `vim.F.if_nil`

glepnir commented on runtime/lua/vim/lsp/document_highlight.lua:
```diff
@@ -0,0 +1,502 @@
+---Move to a document highlight
+---@param opts vim.lsp.document_highlight.JumpOpts
+function M.jump(opts)
+  vim.validate('opts', opts, 'table')
+  vim.validate('count', opts.count, 'number')
```
also there just call once.

justinmk commented on runtime/doc/lsp.txt:
```diff
@@ -1941,6 +1924,23 @@ on_publish_diagnostics({_}, {params}, {ctx})
+==============================================================================
+Lua module: vim.lsp.document_highlight                *lsp-document_highlight*
+
+jump({opts})                               *vim.lsp.document_highlight.jump()*
```
what does `jump({})` with no options do?

justinmk commented on runtime/lua/vim/lsp/util.lua:
```diff
@@ -1735,18 +1735,22 @@ do --[[ References ]]
   ---@param position_encoding 'utf-8'|'utf-16'|'utf-32'
   ---@see https://microsoft.github.io/language-server-protocol/specification/#textDocumentContentChangeEvent
   function M.buf_highlight_references(bufnr, references, position_encoding)
+    vim.deprecate('vim.lsp.util.buf_highlight_references()', 'nil', '0.12')
```
how does one "show a list of doc highlights" without this?

justinmk commented on test/functional/plugin/lsp/document_highlight_spec.lua:
```diff
@@ -0,0 +1,183 @@
+        vim.lsp.document_highlight.jump({ count = 999 })
+      end)
+      screen:expect([[
+  {1:global} = _G                                          |
```
expect dedents the screen, so this could be indented more nicely :)

justinmk commented on runtime/lua/vim/lsp/buf.lua:
```diff
@@ -1059,13 +1059,14 @@ end
 ---         |hl-LspReferenceText|
 ---         |hl-LspReferenceRead|
 ---         |hl-LspReferenceWrite|
```
does the new enable() need some parts of this doc ?

ofseed commented:
I have rechecked most of the logic in this PR, and it is now significantly different from the earliest version (nearly a year ago). I believe it is ready to be merged/reviewed.

justinmk commented on runtime/lua/vim/lsp/document_highlight.lua:
```diff
@@ -0,0 +1,430 @@
+
+local namespace = api.nvim_create_namespace('nvim.lsp.document_highlight')
+api.nvim_set_decoration_provider(namespace, {
+  on_win = function(_, _, bufnr, toprow, botrow)
```
generally want to avoid setting up global handlers on module-load. should this be in enable() somewhere ?

justinmk reviewed (approved):
Looks great, needs conflict resolution though

mfussenegger reviewed (commented):
What's a bit of a bummer is that this is taking away the possibility for users to define there own trigger for the document highlight.

Also - does the window decoration code have a performance advantage for the document highlights? There are usually not too many highlight references and the logic of `buf_highlight_references` was rather simple, compared to what we have now. (40 lines vs. 430)

justinmk commented on runtime/doc/news.txt:
```diff
@@ -216,6 +216,9 @@ LSP
   https://microsoft.github.io/language-server-protocol/specification/#textDocument_linkedEditingRange
 • Support for related documents in pull diagnostics:
   https://microsoft.github.io/language-server-protocol/specifications/specification-current/#relatedFullDocumentDiagnosticReport
+• Re-implemented support for `textDocument/documentHighlight
```
is it enabled by default? if so should mention it here.

ofseed committed:
feat(lsp)!: deprecate `vim.lsp.buf.document_highlight()` related APIs

ofseed committed:
feat(lsp): reimplement LSP document highlight with the new API style

ofseed committed:
wip

ofseed committed:
docs(lua): mark `vim.Pos.Optional` as optional

ofseed committed:
feat(lua): conversion between cursor positions
